src/matrix/room/timeline/Timeline.js

```diff
@@ -236,7 +254,111 @@ export class Timeline {
+         * This code is specific to entries fetched from HomeServer i.e NonPersistedEventEntry
+         */
+        for (const entry of entries) {
+            const relatedEntry = this._contextEntriesNotInTimeline.get(entry.relatedEventId);
+            if (relatedEntry?.addRelation(entry)) {
+                // update other entries for which this entry is a context entry
+                relatedEntry.contextForEntries?.forEach(e => {
+                    this._remoteEntries.findAndUpdate(te => te.id === e.id, () => { return { reply: relatedEntry }; });
+                });
+            }
+        }
+    }
+
+    /**
+     * If an event we had to fetch from hs/storage is now in the timeline (for eg, due to gap fill),
+     * remove the event from _contextEntriesNotInTimeline since it is now in remoteEntries
+     */
+    _moveEntryToRemoteEntries(entries) {
+        for (const entry of entries) {
+            const fetchedEntry = this._contextEntriesNotInTimeline.get(entry.id);
+            if (fetchedEntry) {
+                fetchedEntry.contextForEntries.forEach(e => e.setContextEntry(entry));
+                entry.updateFrom(fetchedEntry);
+                this._contextEntriesNotInTimeline.delete(entry.id);
+            }
+        }
+    }
+
+    _updateEntry(entry) {
+        const txnId = entry.isPending ? entry.id : null;
+        const eventId = entry.isPending ? null : entry.id;
+        this._findAndUpdateEntryById(txnId, eventId, () => true);
+    }
+
+    /**
+     * For each entry in entries, this method associates a context-entry (if needed) to it.
+     * The context-entry is fetched using the following strategies (in the same order as given):
+     * - timeline
+     * - storage
+     * - homeserver
+     * @param {EventEntry[]} entries 
+     */
+    async _loadContextEntriesWhereNeeded(entries) {
+        const entriesNeedingContext = entries.filter(e => !!e.contextEventId);
+        for (const entry of entriesNeedingContext) {
+            const id = entry.contextEventId;
+            let contextEvent = this._getTrackedEntry(id);
+            if (!contextEvent) {
+                contextEvent = await this._getEventFromStorage(id) ?? await this._getEventFromHomeserver(id);
+                if (contextEvent) {
+                    // this entry was created from storage/hs, so it's not tracked by remoteEntries
+                    // we track them here so that we can update reply previews later
+                    this._contextEntriesNotInTimeline.set(id, contextEvent);
+                }
+            }
+            if (contextEvent) {
+                entry.setContextEntry(contextEvent);
+                this._updateEntry(entry);
+            }
+        }
+    }
+
+    /**
+     * Fetches an entry with the given event-id from remoteEntries or contextEntriesNotInTimeline.
+     * @param {string} eventId event-id of the entry
+     * @returns entry if found, undefined otherwise
+     */
+    _getTrackedEntry(eventId) {
```

Comment:
Still wonder if there is a less generic name for this ... what about `_findLoadedEventById`?